File: src/main/java/com/instaimg/crawl/json/CustomJsonParser.java
package com.instaimg.crawl.json;

import org.json.simple.JSONArray;
import org.json.simple.JSONObject;

import java.util.ArrayList;
import java.util.List;

import static com.instaimg.crawl.json.JsonAttr.MULTI_IMG_START_KEY;
import static com.instaimg.crawl.json.JsonAttr.SINGLE_IMG_START_KEY;

public class CustomJsonParser {
    public static List<String> getImageUrl(JSONObject jsonObject) {
        JSONArray jsonArray = (JSONArray) jsonObject.get("items");
        List<String> imgURLs = new ArrayList<>();

        for (Object obj : jsonArray) {
            JSONObject item = (JSONObject) obj;
            if (item.containsKey(SINGLE_IMG_START_KEY.getKey())) {
                imgURLs.add(getSingleImgUrl(item));
            } else if (item.containsKey(MULTI_IMG_START_KEY.getKey())) {
                addMultiImgUrl(item, imgURLs);
            }
        }
        return imgURLs;
    }

    private static String getSingleImgUrl(JSONObject item) {
        JSONObject image_version = (JSONObject) item.get(SINGLE_IMG_START_KEY.getKey());
        JSONArray candidates = (JSONArray) image_version.get("candidates");
        JSONObject candidatesObj = (JSONObject) candidates.get(0);
        return String.valueOf(candidatesObj.get("url"));
    }

    private static void addMultiImgUrl(JSONObject item, List<String> imgURLs) {
        JSONArray carousel_media = (JSONArray) item.get(MULTI_IMG_START_KEY.getKey());
        for (Object obj : carousel_media) {
            JSONObject jsonObj = (JSONObject) obj;
            JSONObject image_version = (JSONObject) jsonObj.get(SINGLE_IMG_START_KEY.getKey());
            JSONArray candidates = (JSONArray) image_version.get("candidates");
            JSONObject candidatesObj = (JSONObject) candidates.get(0);
            String url = String.valueOf(candidatesObj.get("url"));
            imgURLs.add(url);
        }
    }
}

File: src/main/kotlin/com/instaimg/crawl/json/CustomJsonParser.kt
package com.instaimg.crawl.json

import com.instaimg.crawl.json.JsonAttr.MULTI_IMG_START_KEY
import com.instaimg.crawl.json.JsonAttr.SINGLE_IMG_START_KEY
import org.json.simple.JSONArray
import org.json.simple.JSONObject

object CustomJsonParser {

    fun getImageUrl(jsonObject: JSONObject): List<String> =
        (jsonObject["items"] as JSONArray).flatMap { obj ->
            val item = obj as JSONObject
            when {
                item.containsKey(SINGLE_IMG_START_KEY.key) -> listOf(extractFirstCandidateUrl(item, SINGLE_IMG_START_KEY))
                item.containsKey(MULTI_IMG_START_KEY.key) -> extractCarouselUrls(item)
                else -> emptyList()
            }
        }

    private fun extractCarouselUrls(item: JSONObject): List<String> =
        (item[MULTI_IMG_START_KEY.key] as JSONArray).map { obj ->
            extractFirstCandidateUrl(obj as JSONObject, SINGLE_IMG_START_KEY)
        }

    private fun extractFirstCandidateUrl(item: JSONObject, attr: JsonAttr): String {
        val imageVersion = item[attr.key] as JSONObject
        val candidates = imageVersion["candidates"] as JSONArray
        return (candidates[0] as JSONObject)["url"].toString()
    }
}
